designs/documents.md

```diff
@@ -0,0 +1,424 @@
+`Document` type may not include this functionality, but for Python it is a huge
+ease-of-use improvement for those that don't need the extra saftey and
+functionality of the full `Document` type.
+
+Currently the base `Document` implementation lazily converts between the two
+representations and caches the results. This is done for the sake of speed, but
+has an unfortunate space cost. This behavior is not part of the API, however,
+and MAY be changed without backwards compatibility implications.
+
+```python
+>>> document = Document({"foo": "bar"})
+>>> document.shape_type
+ShapeType.DOCUMENT
+>>> document.as_value()
+{"foo": "bar"}
+>>> document["foo"]
+Document(value="bar")
+```
+
+#### Structures and Unions
+
+Smithy structures and unions are generated into their own individual classes
+in Python, and so it isn't possible to have a unique accessor for each of them.
+Instead, `Document`s have a generic `as_shape` method that supports converting
+the `Document` to any `DeserialzeableShape`.
+
+Similarly, a `from_shape` method allows for converting any `SerializeableShape`
+into a document. When a `Document` is constructed this way, it will retain the
+schema of the shape it was constructed from.
+
+```python
+@dataclass
+class ExampleStruct:
+    foo: str
+    bar: str | None = None
+
+    @classmethod
+    def deserialize(cls, deserializer: ShapeDeserializer) -> Self:
+        ...
+
+document = Document({"foo": "spam"})
+print(document.as_shape(ExampleStruct))
+# ExampleStruct(foo="spam")
+
+structure = ExampleStruct(foo="spam", bar="eggs")
+print(Document.from_shape(structure))
+# Document(
+#     value={"foo": Document(value="spam"), "bar": Document(value="eggs")},
+#     schema=EXAMPLE_STRUCT_SCHEMA,
+# )
+```
+
+These methods use the `serialize` and `deserialize` methods under the hood,
+providing private `ShapeSerializer`s and `ShapeDeserializer`s to handle the
+heavy lifting.
+
+### Container Methods
+
+`Document`s are often containers, and so to make `Document` usage as natural as
+possible, most of Python's built-in
+[container methods](https://docs.python.org/3/reference/datamodel.html#emulating-container-types)
+will be supported. If the underlying type cannot support that method, an
+exception will be raised. The following list specifies which container methods
+specifically will be supported and what shape types they apply to.
+
+* `__len__` - Valid for lists, maps, structures, and unions.
+* `__getitem__` - Valid for lists, maps, structures, and unions.
+* `get` - Valid for maps, structures, and unions. This is equivalent to `dict`'s
+  `get` method which allows for a default value to be specified.
+* `__setitem__` - Valid for lists, maps, and structures, and unions. Unions and
+  structures may only use keys representing members that are in their schema.
+  Unions may only replace the current value.
+* `__delitem__` - Valid for lists, maps, and structures. Notably not valid for
```

Comment:
Can't unions have additional null members? Should that case be considered here?

Reply:
They may show up that way on the wire, but that's an underlying protocol behavior that should not be exposed.